File: promise/src/test/java/com/voxeet/testpromise/sub_promise/promise/PromiseSubThrowError.java
package com.voxeet.testpromise.sub_promise.promise;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.junit.Before;
import org.junit.Test;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import com.voxeet.promise.Promise;
import com.voxeet.promise.solve.ErrorPromise;
import com.voxeet.promise.solve.PromiseExec;
import com.voxeet.promise.solve.PromiseSolver;
import com.voxeet.promise.solve.Solver;
import com.voxeet.testpromise.utils.AndroidMockUtil;

/**
 * Created by kevinleperf on 06/04/2018.
 */

public class PromiseSubThrowError {

    @Before
    public void setHandler() {
        Promise.setHandler(AndroidMockUtil.mockMainThreadHandler());
    }

    @Test
    public void test() throws InterruptedException {

        final IllegalStateException[] mException = {null};
        final CountDownLatch latch = new CountDownLatch(1);
        final boolean[] equals = {false};

        final Promise<Integer> resolve_later = new Promise<>(new PromiseSolver<Integer>() {
            @Override
            public void onCall(@NonNull final Solver<Integer> solver) {
                try {
                    throw new IllegalStateException();
                } catch (IllegalStateException e) {
                    mException[0] = e;
                    solver.reject(e);
                }
            }
        });

        System.out.println("executing test");
        execute()
                .then(new PromiseExec<Integer, Integer>() {
                    @Override
                    public void onCall(@Nullable final Integer result, @NonNull final Solver<Integer> solver) {
                        solver.resolve(resolve_later);
                    }
                })
                .error(new ErrorPromise() {
                    @Override
                    public void onError(@NonNull Throwable error) {
                        System.out.println("error catched");
                        equals[0] = mException[0] == error;
                        error.printStackTrace();
                        latch.countDown();
                    }
                });

        //6s are enough
        latch.await(6, TimeUnit.SECONDS);
        if (equals[0] != true) {
            throw new IllegalStateException("Expected exception same...");
        }
    }

    private Promise<Integer> execute() {
        return new Promise<>(new PromiseSolver<Integer>() {
            @Override
            public void onCall(@NonNull Solver<Integer> solver) {
                solver.resolve(10);
            }
        });
    }
}

File: promise/src/test/kotlin/com/voxeet/testpromise/subpromise/promise/PromiseSubThrowError.kt
package com.voxeet.testpromise.subpromise.promise

import com.voxeet.promise.HandlerFactory
import com.voxeet.promise.Promise
import com.voxeet.promise.solve.Solver
import com.voxeet.testpromise.mockedhandler
import org.junit.Before
import org.junit.Test
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

class PromiseSubThrowError {
    @Before
    fun setHandler() {
        HandlerFactory.setHandler(mockedhandler())
    }

    @Test
    fun test() {
        val mException = arrayOf<IllegalStateException?>(null)
        val latch = CountDownLatch(1)
        val equals = booleanArrayOf(false)
        val resolve_later = Promise { solver: Solver<Int?> ->
            try {
                throw IllegalStateException()
            } catch (e: IllegalStateException) {
                mException[0] = e
                solver.reject(e)
            }
        }
        println("executing test")
        execute().then { result: Int?, solver: Solver<Int?> ->
            solver.resolve(
                resolve_later
            )
        }.error { error: Throwable ->
            println("error catched")
            equals[0] = mException[0] === error
            error.printStackTrace()
            latch.countDown()
        }

        //6s are enough
        latch.await(6, TimeUnit.SECONDS)
        check(equals[0] == true) { "Expected exception same..." }
    }

    private fun execute(): Promise<Int?> {
        return Promise { solver: Solver<Int?> -> solver.resolve(10) }
    }
}
